lib/elixir/lib/uri.ex

```diff
@@ -977,6 +977,42 @@ defmodule URI do
       %{uri | query: uri.query <> "&" <> query}
     end
   end
+
+  @doc """
+  Appends `path` to the given `uri`.
+
+  Any additional url components in `path` (like query strings) will be removed.
+
+  ## Examples
+
+      iex> URI.append_path(URI.parse("http://example.com"), "my-path") |> URI.to_string()
+      "http://example.com/my-path"
+
+      iex> URI.append_path(URI.parse("http://example.com/foo/?x=1"), "/my-path") |> URI.to_string()
+      "http://example.com/foo/my-path?x=1"
+
+      iex> URI.append_path(URI.parse("http://example.com"), "/my-path?x=1") |> URI.to_string()
+      "http://example.com/my-path"
+  """
+  @doc since: "1.15.0"
+  @spec append_path(t(), binary()) :: t()
+  def append_path(%URI{} = uri, path) when is_binary(path) do
+    current_path = uri.path || ""
+    to_append = parse(path).path
```

Comment:
we need to be careful with `URI.parse(path).path` because it drops everything preceding the path from the url. For example:

```elixir
iex> URI.append_path(URI.parse("https://play.hbomax.com/page/"), "urn:hbo:page") |> URI.to_string()
"https://play.hbomax.com/page/hbo:page"

# it should be:

"https://play.hbomax.com/page/urn:hbo:page"
```

I think we should append stuff verbatim and document as such but yeah, I'm still not sure how we should handle query strings in either the URL or the path to append. One of the doctests,

```
"http://example.com/foo/?x=1" + "/my-path" #=> "http://example.com/foo/my-path?x=1"
```

feels awkward to me but I can't quite explain why exactly. FWIW `URI.merge` (and `:uri_string.resolve/2`) would drop the query string.

Reply:
Correct. I think using parse/1 is actually complicating the solution. I would instead expect a string starting with “/“ to be given. If it starts “//“, you should raise. If it has none, it should raise.

Reply:
> If it starts “//“, you should raise. If it has none, it should raise.

That makes sense. Since it raises, does it need to be `append_path!/2` and then should there be an `append_path/2` that returns a tuple?

Reply:
No. We can always raise on bad input. The bang variant would make sense if "//" would work sometimes but not others.